ruby/phase-1.md

```diff
@@ -0,0 +1,398 @@
+support direct workflow invocations (removing all SDK traces out of the way):
+
+```ruby
+MyWorkflow.start(arg_1, arg_2, keyword_arg: arg_3)
+```
+
+Once a workflow is started all the interactions will be going through the handle object returned to
+the caller. The handle can of course be created from a namespace/workflow_id/run_id combination
+using the `Client#workflow_handle` method. Here the interface of the handle:
+
+```ruby
+class WorkflowHandle
+  def initialize(connection, namespace, workflow_id, run_id = nil)
+
+  def result(await: false, timeout: nil) -> Any
+
+  def describe -> WorkflowExecutionInfo
+
+  def cancel -> void
+
+  def query(name, *args) -> Any
+
+  def signal(name, *args) -> void
+
+  def terminate(reason, details: nil) -> void
+end
+
+```
+
+
+## Worker
+
+The worker interface allows SDK users to register any number of both workflows and activities. These
+will then be polled for on the specified namespaces & task queues and executed. The worker can poll
+on multiple task queues at the same time. These will be determined by the registered activities and
+workflow. If there are any overlaps we'll only create one poller for each unique task queue.
+
+```ruby
+config Temporal::Configuration.new(url: 'localhost:7233')
+worker = Temporal::Worker.new(config)
+
+worker.register_workflow(NewWorkflow)
+worker.register_workflow(NewWorkflow, task_queue: 'tq-3')
+worker.register_activity(AnotherActivity)
+worker.register_activity(AnotherActivity, namespace: 'ns-3')
+
+# Calling Worker#run will block the main thread until receiving SIGTERM or SIGINT
+worker.run
+```
+
+Notes:
+
+- The worker will poll for workflows on `ns-1`/`tq-1` and `ns-1`/`tq-3`
+- And activities on `ns-2`/`tq-2` and `ns-3`/`tq-2`
+- The workflows/activities can of course be defined in any number of different workers
+- The worker is intended to take over the process
+
+### Concurrency model
+
+Ruby can be a tricky language for running concurrent applications. It supports kernel threads,
+however it has a Global Interpreter Lock (GIL) in place which prevents any two threads to be
+executed at the same time (making it effectively single-threaded). The good news is that GIL does
+allow IO-bound instructions to be executed in parallel.
+
+Because of these limitations the most resource efficient solution is a combination of process
+forking and threading. The exact ratio can depend heavily on the application and hardware used. This
+is why it is crucial to allow the Ruby SDK users to configure how many processes/threads they need
+per each worker.
+
+Internally a worker will run each poller in its own thread and will create 2 shared thread pools of
+a given size for executing tasks (one pool for activities and one for workflows). A poller will
+reserve a thread on a thread pool before polling to ensure there's capacity for processing a task.
+
+To support multiple processes we'll have a `Temporal::Crew` class that run a given worker
```

Comment:
Hmmm, I see where you're going with this. Feels like a concern that could be handled by an add-on gem at a later date. Is this really core to the SDK?

Reply:
That is, supervisory frameworks with dead-child reaping and restarts are nice and Erlangy, but feel like a nice-to-have, not a must-have.

Reply:
Getting inspiration from Sidekiq here where they include this option in their paid versions. But you're right and we can probably avoid it and just hint at it in the README (there's probably solutions in place to achieve this kind of result)

Reply:
Agree this shouldn't be a part of the core SDK.

Reply:
No problem, I'll scrap it 👍